### PR TITLE
Add useful methods to PulseBackendConfiguration

### PR DESCRIPTION
**Pull request by lcapelluto**
### Summary
Improve usability of backend config by adding helpful methods to parse configuration info

 - `drive(self, qubit)`: returns the drive channel of `qubit`
 - `measure(self, qubit)`: returns the measure channel of `qubit`
 - `acquire(qubit)`: returns the acquire channel of `qubit`
 - `control(qubit)`: returns the control channel of `qubit` (sends a warning, as well, because this isn't correct, but matches current behavior of PulseChannelSpec)
 - `describe(channel: ControlChannel)`: returns a description of `channel` 

This shall replace `PulseChannelSpec`.

TODO:
- [x] Make an issue for implementing `draw()` feature (drawing the topology of the backend) #3310  -- closed
- [x] Make an issue for implementing `control(qubit)` #3311

### CODE REVIEW

**Comment by lcapelluto** [comment]
Tests failing because visualizer expects reference good images for each backend, and FakeOpenpulse3Q doesn't have it

File: qiskit/providers/models/backendconfiguration.py [hunk 1]
```diff
@@ -300,3 +340,115 @@ def __init__(self, backend_name, backend_version, n_qubits, basis_gates,
                          meas_lo_range=meas_lo_range, dt=dt, dtm=dtm,
                          rep_times=rep_times, meas_kernels=meas_kernels,
                          discriminators=discriminators, **kwargs)
+
+    @property
+    def dt(self) -> float:  # pylint: disable=invalid-name
+        """Time delta between samples on the signal channels in seconds."""
+        warnings.warn("The time delta is now returned in units of [s] rather than [ns].")
```
Comment (lcapelluto) [inline qiskit/providers/models/backendconfiguration.py]: This is worth looking at closely: this is a hard change for the user. Is there a better way to do this? We do consider Pulse to sort of be in beta, but can we make this change easier for them? Same for `dtm`. No other existing behavior changes in this PR.

Comment (taalexander) [reply qiskit/providers/models/backendconfiguration.py]: I think we may have to bite the bullet here and note it in the changelogs.

File: qiskit/providers/models/backendconfiguration.py [hunk 2]
```diff
@@ -300,3 +340,115 @@ def __init__(self, backend_name, backend_version, n_qubits, basis_gates,
...
+        """Time delta between samples on the acquisition channels in seconds."""
+        warnings.warn("The time delta is now returned in units of [s] rather than [ns].")
+        return self._dtm * 1e-9
+
+    @property
+    def sample_rate(self) -> float:
+        """Sample rate of the signal channels in Hz (1/dt)."""
+        return 1.0 / self.dt
+
+    def hamiltonian(self) -> str:
```
Comment (taalexander) [inline qiskit/providers/models/backendconfiguration.py]: In the future this should return a Hamiltonian object. I think for now we should just return the Hamiltonian dictionary as this is the Hamiltonian "object". I believe @DanPuzzuoli will be working on doing a proper implementation moving forward.

File: qiskit/providers/models/backendconfiguration.py [hunk 3]
```diff
@@ -300,3 +340,115 @@ def __init__(self, backend_name, backend_version, n_qubits, basis_gates,
...
+        """
+        return getattr(self, '_hamiltonian', {}).get('h_latex', None)
+
+    def hamiltonian_description(self) -> str:
+        """
+        Return the Hamiltonian description for this device if provided.
+        """
+        return getattr(self, '_hamiltonian', {}).get('description', None)
+
+    def drive(self, qubit: int) -> DriveChannel:
```
Comment (taalexander) [inline qiskit/providers/models/backendconfiguration.py]: For ease of use should this then also be exposed as a method directly on the backend, @ajavadia @dcmckayibm? This seems to make more sense to me.

Comment (taalexander) [inline qiskit/providers/models/backendconfiguration.py]: I think these would be better exposed to a `ChannelGenerator` object that the BackendConfiguration has and then exposes this behavior by calling the underlying object. Right now the object is doing too much. This is nonblocking.

File: qiskit/test/mock/fake_openpulse_2q.py [hunk 4]
```diff
@@ -65,7 +65,10 @@ def __init__(self):
             conditional_latency=[
                 [100, 1000], [1000, 100], [100, 1000],
                 [1000, 100], [100, 1000], [1000, 100]
-            ]
+            ],
+            hamiltonian={
+                'h_latex': "MOCK_HAMILTONIAN_STR"
```
Comment (taalexander) [inline qiskit/test/mock/fake_openpulse_2q.py]: This is not valid, could you please add a valid Hamiltonian which should include `h_str`, `h_latex` and `vars`, `qub`. See this example for more info https://github.com/Qiskit/qiskit-aer/blob/6c2dd4ed7c42c9d6775293d9832b6779fb3abed9/example/pulse_sim.ipynb.

File: releasenotes/notes/extend-backend-configuration-129d7292471fe331.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+---
+features:
+  - |
+    PulseBackendConfiguration (accessed normally as backend.configuration())
+    has been extended with useful methods to explore its data and the
+    functionality that exists in PulseChannelSpec. PulseChannelSpec will be 
+    deprecated in the future.
+
+    Example:
+        backend = provider.get_backend(backend_name)
```
Comment (taalexander) [inline releasenotes/notes/extend-backend-configuration-129d7292471fe331.yaml]: Could you also please include the `dt` unit change as a note?

File: qiskit/providers/models/backendconfiguration.py [hunk 6]
```diff
@@ -243,53 +263,73 @@ def __init__(self, backend_name, backend_version, n_qubits, basis_gates,
...
         self.meas_lo_range = meas_lo_range
-        self.dt = dt  # pylint: disable=invalid-name
-        self.dtm = dtm
+        self._dt = dt  # pylint: disable=invalid-name
+        self._dtm = dtm
         self.rep_times = rep_times
         self.meas_kernels = meas_kernels
         self.discriminators = discriminators
+        if 'hamiltonian' in kwargs:
+            self._hamiltonian = kwargs.pop('hamiltonian')
```
Comment (1ucian0) [inline qiskit/providers/models/backendconfiguration.py]: The way to have optional parameters is with `hamiltonian=None` in the `__init__` definition. I thinks that is clearer.

File: test/python/providers/test_backendconfiguration.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+class TestBackendConfiguration(QiskitTestCase):
+    """Test the methods on the BackendConfiguration class."""
+
+    def setUp(self):
+        self.provider = FakeProvider()
+        self.config = self.provider.get_backend('fake_openpulse_2q').configuration()
+
+    def test_simple_config(self):
+        """Test the most basic getters."""
+        self.assertEqual(self.config._dt, self.config.dt / 1.e-9)
```
Comment (1ucian0) [inline test/python/providers/test_backendconfiguration.py]: I dont think internal attributes should be checked.

File: qiskit/providers/models/backendconfiguration.py [hunk 8]
```diff
@@ -243,53 +263,73 @@ def __init__(self, backend_name, backend_version, n_qubits, basis_gates,
...
+            open_pulse: backend supports open pulse.
+            memory: backend supports memory.
+            max_shots: maximum number of shots supported.
+            n_uchannels: Number of u-channels.
+            u_channel_lo: U-channel relationship on device los.
+            meas_levels: Supported measurement levels.
+            qubit_lo_range: Qubit lo ranges for each qubit with form (min, max) in GHz.
+            meas_lo_range: Measurement lo ranges for each qubit with form (min, max) in GHz.
+            dt: Qubit drive channel timestep in nanoseconds.
+            dtm: Measurement drive channel timestep in nanoseconds.
```
Comment (1ucian0) [inline qiskit/providers/models/backendconfiguration.py]: ?
```suggestion
            dtm: Measurement drive channel timestep in seconds.
```

File: qiskit/providers/models/backendconfiguration.py [hunk 9]
```diff
@@ -243,53 +263,73 @@ def __init__(self, backend_name, backend_version, n_qubits, basis_gates,
...
+            conditional: backend supports conditional operations.
+            open_pulse: backend supports open pulse.
+            memory: backend supports memory.
+            max_shots: maximum number of shots supported.
+            n_uchannels: Number of u-channels.
+            u_channel_lo: U-channel relationship on device los.
+            meas_levels: Supported measurement levels.
+            qubit_lo_range: Qubit lo ranges for each qubit with form (min, max) in GHz.
+            meas_lo_range: Measurement lo ranges for each qubit with form (min, max) in GHz.
+            dt: Qubit drive channel timestep in nanoseconds.
```
Comment (1ucian0) [inline qiskit/providers/models/backendconfiguration.py]: ?
```suggestion
            dt: Qubit drive channel timestep in seconds.
```

File: qiskit/providers/models/backendconfiguration.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/providers/models/backendconfiguration.py]: I agree, not intuitive at all. I'm not sure if this is something I can fix in terra, but I will look

[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/providers/models/backendconfiguration.py]: AFAIK, the individual providers are responsible for this. It's now set in place by the OpenPulse Specification. I can't change it (maybe on larger timescale I can)

File: qiskit/providers/models/backendconfiguration.py (hunk 3 above)
Comment (taalexander) [reply qiskit/providers/models/backendconfiguration.py]: I agree with them being too simple to be even really useful. The question comes back to how do we get u-channels to users for what they want to do, and always comes back to either the u-channel lo or the Hamiltonian.

Comment (lcapelluto) [reply qiskit/providers/models/backendconfiguration.py]: So, you think we should fully deprecate these methods? Just keep `control`?

**Pull request merged**